v3-roadmap-and-design.md

```diff
@@ -0,0 +1,425 @@
+
+## Goals
+
+- Provide a complete implementation of Zarr V3 through the Zarr-Python API
+- Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)
+- Clear the way for exciting extensions / ZEPs (i.e. [sharding](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html), [variable chunking](https://zarr.dev/zeps/draft/ZEP0003.html), etc.)
+- Provide a developer API that can be used to implement and register V3 extensions
+- Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
+- Clean up the internal and user facing APIs 
+- Improve code quality and robustness (e.g. achieve 100% type hint coverage)
+
+## Examples of what 3.0 will enable?
+1. Reading and writing V3 spec-compliant groups and arrays
+2. V3 extensions including sharding and variable chunking.
+3. Consolidation of writes (e.g. array metadata + attributes concurrently)
+4. Improved concurrency when creating/reading/writing to stores (imagine a `create_hierarchy(zarr_objects)` function).
+5. User-developed extensions (e.g. storage-transformers) can be registered with Zarr-Python at runtime
+
+## Non-goals (of this document)
+
+- Implementation of any Zarr V3 extensions
+- Major revisions to the Zarr V3 spec
+
+## Requirements
+
+1. Read and write spec compliant V2 and V3 data
+2. Limit unnecessary traffic to/from the store
+3. Cleanly define the Array/Group/Store abstractions
+4. Cleanly define how V2 will be supported going forward
+5. Provide a clear roadmap to help users upgrade to 3.0
+6. Developer tools / hooks for registering extensions
+
+## Design
+
+### Async API
+
+Zarr-Python is an IO library. As such, supporting concurrent action against the storage layer is critical to achieving acceptable performance. The Zarr-Python 2 was not designed with asynchronous computation in mind and as a result has struggled to effectively leverage the benefits of concurrency. At one point, `getitems` and `setitems` support was added to the Zarr store model but that is not leveraged throughout the API. 
+
+With Zarr-Python 3.0, we have the opportunity to revisit this design. The proposal here is as follows:
+
+1. The `Store` interface will be entirely async. 
+2. On top of the async `Store` interface, we will provide an  `AsyncArray` and `AsyncGroup` interface.
+3. Finally, the primary user facing API will be synchronous `Array` and `Group` classes that wrap the async equivalents.
```

Comment:
> We probably also need some concept of "batch" and/or "queue" to allow files to accumulate or stream to the async IO.

I think tensorstore achieves something like this via [transactions](https://google.github.io/tensorstore/python/api/tensorstore.Transaction.html), it might be worth copying that API

Reply:
We have been discussing something similar that could exist at (or near) the store layer. 

```python

store = MyStore()

with store.transaction() as ts:
    array_foo.create(..., store=ts)
    array_bar.create(..., store=ts)
```

This is going to necessitate some thought around the state of the store before the transaction but would enable matching multiple writes in the form of a write-cache.

Reply:
Wouldn't it be possible for users to use the AsyncArray and AsyncGroup directly, if concurrency is important to them?